# The Junction Usage Model (JUM)
- JUM is a tool that offers a completely annotation-free approach to analyze alternative pre-mRNA splicing patterns that are specific to a tissue or sample of interest from RNA-seq datasets.
    * Stable performance package for V3.0.0: <https://github.com/qqwang-berkeley/JUM/releases/tag/v3.0.0>
    * Complete manual for V3.0.0 and V2.0.2: <https://github.com/qqwang-berkeley/JUM/wiki/A-Hitchhiker's-guide-to-JUM-(version-2.0.2--and-3.0.0)>
    * Legacy release for V2.0.2: <https://github.com/qqwang-berkeley/JUM/releases/tag/JUM_v2.0.2>
    * For differential AS analysis among multiple experimental conditions see: https://github.com/qqwang-berkeley/JUM/wiki/6.-Running-JUM-for-differential-AS-analysis-among-multiple-experimental-conditions-(multiple-tissues,-patients,-time-series-experiments-etc.)

- JUM reports alternative splicing events in six categories: cassette exon, mutually exclusive exons, alternative 5' splice site, alternative 3' splice site, intron retention, Composite patterns

- JUM does not depend on any prior knowledge of a transcriptome annotation nor known libraries of AS events.

- The JUM method is published: <https://www.ncbi.nlm.nih.gov/pubmed/30104386>
The simulated RNA-seq datasets and ground truth (true AS genes and non-AS genes) described in the paper are deposited in GEO with accession number GSE118193.
- Feel free to post any questions or comments on the Google group for JUM users: <https://groups.google.com/forum/#!forum/jum-users>


* Legacy manual for V2.0 and V2.0.1: <https://github.com/qqwang-berkeley/JUM/wiki/A-Hitchhiker's-guide-to-JUM-(version-2.0-and-version-2.0.1)>
* Legacy manual for V1.3.12: <https://github.com/qqwang-berkeley/JUM/wiki/A-Hitchhiker's-guide-to-JUM-(version-1.3.12)>
